proposals/NNNN-input-semantics.md

```diff
@@ -0,0 +1,245 @@
+
+## Proposal
+
+SPIR-V and DXIL semantic lowering is very different. SPIR-V respects the
+parameters/field order, while DXIL packs by following an extensive set of
+rules. The System vs User semantics handling is also divergent.
+
+This means we will be able to share the Sema checks, but will have to build
+two distinct paths during codegen.
+
+### Sema
+
+Type check for semantics is currently handled in `SemaDeclAttr.cpp`. \
+Iteration is done on each declaration, and if a semantic attribute is present,
+the type is checked.
+Each declaration being handled independently, this method does not support
+inherited/shadowed semantic attributes.
+
+Sema checks are divided into three parts:
+ - check for type compatibility between the variable and the semantic.
+ - check for semantic duplication.
+ - check for invalid system semantic usage depending on shader stage.
+
+Proposition is to remove the type validation from the `SemaDeclAttr` and
+move it later into SemaHLSL, with the other checks.
+Idea is we need to have the inheritance rules to check types, so we should
+avoid duplicating this logic in two places.
+
+The pseudo-code for this check should be:
+
+```cpp
+  void checkSemantic(std::unordered_set<HLSLAnnotationAttr> &UsedSemantics,
+                DeclaratorDecl *Decl,
+                HLSLAnnotationAttr *InheritedSemantic = nullptr) {
+
+    HLSLAnnotationAttr *Semantic = InheritedSemantic ? InheritedSemantic : Decl->get<HLSLAnnotationAttr>();
+    RecordDecl *RD = dyn_cast<RecordDecl>(Decl);
+
+    // Case 1: type is a scalar, and we have a semantic. End case.
+    if (Semantic && !RD) {
+      if (UsedSemantics.contains(Semantic) && !InheritedSemantic)
+        Fail("Explicit semantic duplication", Decl->getLocation())
+
+      UsedSemantics.insert(Semantic)
+      diagnoseSemanticType(Decl, Semantic);
+      diagnoseSemanticEnvironment(Decl, Context.ShaderEnv);
+      return;
+    }
+
+    // Case 2: type is scalar, but we have no semantic: error
+    if (!RD)
+      Fail("Missing semantic", Decl->getLocation());
+
+    // Case 3: it's a struct. Simply recurse, optionnally inherit semantic.
+    if (RecordDecl *RD = dyn_cast<RecordDecl>(Decl)) {
+      for (FieldDecl *FD : Decl->asRecordDecl()->getFields())
+        checkSemantic(UsedSemantics, FD, Semantic);
+    }
+  }
+
+  ...
+
+  std::unordered_set<clang::HLSLAnnotationAttr> UsedSemantics;
+  for (ParmVarDecl Decl : entrypoint->getParams()) {
+    checkSemantic(UsedSemantics, Decl, /* InheritedSemantic= */ nullptr);
+  }
+```
+
+At this point, we are guaranteed to have only valid and unique semantics, as
+well as valid types.
+
+### CodeGen
+
+DXIL and SPIR-V codegen will be very different, but the flattening/inheritance
+bit can be shared.
+
+The proposal is to provide a sorted list in `CGHLSLRuntime`:
+
+```cpp
+struct SemanticIO {
+  // The active semantic for this scalar/field.
+  HLSLAnnotationAttr *Semantic;
+
+  // Info about this field/scalar.
+  DeclaratorDecl *Decl;
```

Comment:
This would be useful to handle the `[[vk::location(X)]]` attribute. Would be handled later, on top of this.